controllers/storagecluster/server.go

```diff
@@ -0,0 +1,190 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/multierr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+)
+
+var (
+	ServerImage = "quay.io/sp1098/ocs-operator:consumer-api"
+)
+
+type ocsProviderServer struct{}
+
+func (obj *ocsProviderServer) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	if !instance.Spec.AllowRemoteStorageConsumers {
+		r.Log.Info("nigoyal Spec.AllowRemoteStorageConsumers is disabled")
+		return obj.ensureDeleted(r, instance)
+	}
+
+	r.Log.Info("nigoyal Spec.AllowRemoteStorageConsumers is enabled creating resources")
+
+	desiredServerService := GetProviderServerService(instance)
+	desiredServerDeployment := GetProviderServerDeployment(instance)
+
+	var finalErr error
+
+	for _, obj := range []client.Object{
+		desiredServerDeployment,
+		desiredServerService,
+	} {
+		_, err := controllerutil.CreateOrUpdate(
+			context.TODO(), r.Client, obj,
+			func() error { return nil },
+		)
+
+		if err != nil && !errors.IsAlreadyExists(err) {
+			r.Log.Error(err, "nigoyal Failed to create resource", "Kind", obj.GetObjectKind(), "Name", obj.GetName())
+			multierr.AppendInto(&finalErr, err)
+			continue
+		}
+
+		r.Log.Info("nigoyal Resource created successfully", "Kind", obj.GetObjectKind(), "Name", obj.GetName())
+	}
+
+	if finalErr != nil {
+		return finalErr
+	}
+
+	if err := obj.isDeploymentReady(r, instance); err != nil {
```

Comment:
Why do you need to pass `r` and `instance` into this? I would imagine you should be able to just pass the Deployment you already have and do the calculation.

Reply:
I don't have the actual copy of the deployment, I have only the desired deployment. r is having a client which will help fetch the actual deployment with the help of an instance and verify its status.

Reply:
I don't mean the desired Deployment, I mean the actual Deployment. That's what `resources` here gets set to, right? https://github.com/red-hat-storage/ocs-operator/pull/1421/files#diff-85e662b500fee0a341fbae01f02cd05d31663d6f6e355ab6536eb8b2ca0ea38fR53

Reply:
done